### PR TITLE
Enable internal use of `std::allocator` related functionality

### PR DESCRIPTION
**Pull request by miscco**
This ports all of the allocator related facilities to be usable internally

### CODE REVIEW

**Comment by harrism** [comment]
This is massive Perhaps the PR description should explain more? Also type in title ("interal").

File: libcudacxx/include/cuda/std/detail/libcxx/include/__config [hunk 1]
```diff
@@ -1626,7 +1626,8 @@ typedef unsigned int   char32_t;
 
 #if !defined(_LIBCUDACXX_HAS_NO_ALIGNED_ALLOCATION) && \
     (defined(_LIBCUDACXX_HAS_NO_LIBRARY_ALIGNED_ALLOCATION) || \
-    (!defined(__cpp_aligned_new) || __cpp_aligned_new < 201606))
+    (!defined(__cpp_aligned_new) || __cpp_aligned_new < 201606)) \
+  || defined(__cuda_std__) // FIXME: Properly handle aligned allocations
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__config:1630]: The alignment of this line seems wonky.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__memory/allocation_guard.h [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  {
+    if (__ptr_ != nullptr)
+    {
+      allocator_traits<_Alloc>::deallocate(__alloc_, __ptr_, __n_);
+    }
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI _LIBCUDACXX_INLINE_VISIBILITY _CCCL_CONSTEXPR_CXX14 _Pointer
+  __release_ptr() noexcept
+  { // not called __release() because it's a keyword in objective-c++
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__memory/allocation_guard.h:78]: Gotta love objc++.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__new/allocate.h [hunk 3]
```diff
@@ -62,7 +62,7 @@ inline _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __is_overaligned_for_new(siz
 template <class... _Args>
 _LIBCUDACXX_INLINE_VISIBILITY void* __libcpp_operator_new(_Args... __args)
 {
-#if __has_builtin(__builtin_operator_new) && __has_builtin(__builtin_operator_delete)
+#if 0 && __has_builtin(__builtin_operator_new) && __has_builtin(__builtin_operator_delete)
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__new/allocate.h:65]: What's the actual thing that makes us disable this? I'd rather have that spelled out, even if it's currently universal, than dropping in `false`. Or at least a comment as to why this is disabled would be appreciated.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__new/allocate.h [hunk 4]
```diff
@@ -72,7 +72,7 @@ _LIBCUDACXX_INLINE_VISIBILITY void* __libcpp_operator_new(_Args... __args)
 template <class... _Args>
 _LIBCUDACXX_INLINE_VISIBILITY void __libcpp_operator_delete(_Args... __args)
 {
-#if __has_builtin(__builtin_operator_new) && __has_builtin(__builtin_operator_delete)
+#if 0 && __has_builtin(__builtin_operator_new) && __has_builtin(__builtin_operator_delete)
```
Comment (griwes) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__new/allocate.h:75]: Ditto.

File: libcudacxx/test/libcudacxx/std/utilities/memory/allocator.traits/allocator.traits.members/select_on_container_copy_construction.pass.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        const A<int> a(0);
+        assert(cuda::std::allocator_traits<A<int> >::select_on_container_copy_construction(a).id == 0);
+    }
+    {
+        typedef IncompleteHolder* VT;
+        typedef A<VT> Alloc;
+        Alloc a;
+        assert(cuda::std::allocator_traits<Alloc>::select_on_container_copy_construction(a).id == 0);
+    }
+#if TEST_STD_VER >= 11
```
Comment (griwes) [inline libcudacxx/test/libcudacxx/std/utilities/memory/allocator.traits/allocator.traits.members/select_on_container_copy_construction.pass.cpp:66]: Clean all of these up (not only do we no longer support 03, but also these are actually wrong with how our TEST_STD_VER works :'D).

**Review by griwes — CHANGES_REQUESTED** [review]
Looks good overall, the cleanup of C++11 checks in the tests is what I'm after the most by requesting changes here.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__config [hunk 6]
```diff
@@ -1626,7 +1626,8 @@ typedef unsigned int   char32_t;
...
 #  define _LIBCUDACXX_HAS_NO_ALIGNED_ALLOCATION
 #endif
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__config:1633]: I used the formatter to give it a better look.

god I hate what that does to macros

**Comment by miscco** [comment]
@griwes can you give this another look?

**Review by griwes — APPROVED** [review]

**Pull request merged**